pkg/apis/resource/structured/namedresources/types.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namedresources
+
+import "k8s.io/apimachinery/pkg/api/resource"
+
+// Resources is used in NodeResourceModel.
+type Resources struct {
+	Instances []Instance
+}
+
+type Instance struct {
+	// Name is unique identifier among all resource instances managed by
+	// the driver on the node. It must be a DNS subdomain.
+	Name string
+
+	// Attributes defines the attributes of this resource instance.
+	// The name of each attribute must be unique.
+	Attributes []Attribute
+}
+
+type Attribute struct {
+	// Name is unique identifier among all resource instances managed by
+	// the driver on the node. It must be a DNS subdomain.
+	Name string
+
+	AttributeValue
+}
+
+// AttributeValue must have one and only one field set.
+type AttributeValue struct {
+	QuantityValue    *resource.Quantity
+	BoolValue        *bool
+	IntValue         *int64
+	IntSliceValue    *IntSlice
```

Comment:
I first tried with just `*[]int64` here, but code generation stumbled over that, therefore I am now using the `IntSlice` struct as wrapper around a `[]int64`.

Reply:
I am seeing random failures during round-trip testing when the fuzzer ends up populating this field:
```
panic: reflect: reflect.Value.Set using unaddressable value [recovered]
	panic: reflect: reflect.Value.Set using unaddressable value

goroutine 433 [running]:
testing.tRunner.func1.2({0x2efbe20, 0xc000c098c0})
	/nvme/gopath/go-1.21.0/src/testing/testing.go:1545 +0x366
testing.tRunner.func1()
	/nvme/gopath/go-1.21.0/src/testing/testing.go:1548 +0x630
panic({0x2efbe20?, 0xc000c098c0?})
	/nvme/gopath/go-1.21.0/src/runtime/panic.go:920 +0x270
reflect.flag.mustBeAssignableSlow(0x15)
	/nvme/gopath/go-1.21.0/src/reflect/value.go:272 +0x8c
reflect.flag.mustBeAssignable(...)
	/nvme/gopath/go-1.21.0/src/reflect/value.go:259
reflect.Value.Set({0x300e780?, 0xc0013a72f0?, 0x4499c0?}, {0x300e780?, 0x0?, 0x4af409?})
	/nvme/gopath/go-1.21.0/src/reflect/value.go:2254 +0x73
k8s.io/apimachinery/pkg/runtime.sliceToUnstructured({0x2ecc2a0?, 0xc00112c498?, 0xc001446760?}, {0x300e780?, 0xc0013a72f0?, 0xc0013a7140?})
	/nvme/gopath/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/converter.go:738 +0x8f0
...
```

```
(dlv) p item.NodeResourceModel.NamedResources.Instances[0].Attributes[0].AttributeValue
k8s.io/api/resource/structured/namedresources/v1alpha1.AttributeValue {
	QuantityValue: *k8s.io/apimachinery/pkg/api/resource.Quantity {
		i: (*"k8s.io/apimachinery/pkg/api/resource.int64Amount")(0xc001430040),
		d: (*"k8s.io/apimachinery/pkg/api/resource.infDecAmount")(0xc001430050),
		s: "",
		Format: "DecimalExponent",},
	BoolValue: *false,
	IntValue: *int64 nil,
	IntSliceValue: *k8s.io/api/resource/structured/namedresources/v1alpha1.Int64Slice {
		Ints: []int64 len: 0, cap: 0, nil,},
	StringValue: *"ųǪ",
	StringSliceValue: *k8s.io/api/resource/structured/namedresources/v1alpha1.StringSlice nil,}
```

I need to investigate this further...

Reply:
>I first tried with just *[]int64 here, but code generation stumbled over that

It generates invalid code:
```
	if len(m.IntSliceValue) > 0 {

invalid argument: m.IntSliceValue (variable of type *[]int64) for len
```

Reply:
> I need to investigate this further...

I have a simple reproduce in the corresponding unit test, see https://github.com/kubernetes/kubernetes/issues/123556

I suppose in this PR the quickest path forward is to remove the support for string and int slices.

@klueska: WDYT?

Reply:
Why can't it just be `[]int64`?  There is still a distinction between `nil` and `[]int64{}` here.

Reply:
Never mind - I found my mistake. One cannot inline a slice in a struct in JSON...

Reply:
Suppose someone writes:
```
attributes:
  boolValue: true
  intSliceValue: null
```

That gets decoded into
```
AttributeValue{
   BoolValue: ptr.To(true),
   intSliceValue: nil, 
}
```

It's redundant, but that is okay. When re-encoding, we should use use `omitempty`, so it becomes:
```
attributes:
  boolValue: true
```

Now consider if we do the same with an empty slice:
```
attributes:
  intSliceValue:
```

This decodes into
```
AttributeValue {
   IntSliceValue: []int64{},
}
```

That's still okay.

But encoding it with `omitempty` yields:
```
attributes:{}
```

and that results in an empty `AttributeValue{}`, which is not okay.

We could drop the `omitempty` and always encode the `intSliceValue` and `stringSliceValue`, but that seems ugly and inconsistent to me because the other pointers don't get encoded.

Reply:
Got it -- makes sense. Thanks for the explanation.

Reply:
> So the issue is that omitempty treats an empty slice and a nil slice the same?

Yes.

> give it a custom marshaler

That could solve it for JSON, but then there's still protobuf. I don't know how that behaves. The pointer to a struct is not going to result in surprises.

Reply:
proto encoding does not distinguish between null and zero-length lists, regardless of omitempty. It's really annoying.

https://github.com/kubernetes/kubernetes/issues/44593

Reply:
So let's keep the current "pointer to struct" approach because that doesn't need any complex tricks.